ref(Alg): Modularize algorithms and improved progress bar

Simpag opened the pull request:
Modularization of algorithms into `initialize`, `step` and optional `finalize` methods. Initialization arguments to agents `x0` variable is also added. The progress bar has been updated to include more information. This includes which trial is currently active, optional speed- and progesss step logging. When the progress step logging is active the speed is normalized by the `progress step` parameter as closely as possible. 

Respective documentation updates have been applied.

closes #190 #180 #191

nicola-bastianello commented on decent_bench/distributed_algorithms.py:
```diff
@@ -41,33 +104,33 @@ class DGD(Algorithm):
+        for i in network.agents():
+            if self.x0 is None:
+                self.x0 = iop.zeros(framework=i.cost.framework, shape=i.cost.shape, device=i.cost.device)
+            self.x0 = iop.to_array(self.x0, framework=i.cost.framework, device=i.cost.device)
```
A fix for this could be to do
```
i = network.agents()[0]
if self.x0 is None:
    self.x0 = iop.zeros(framework=i.cost.framework, shape=i.cost.shape, device=i.cost.device)
self.x0 = iop.to_array(self.x0, framework=i.cost.framework, device=i.cost.device)
```
where `i = network.agents()[0]` is changed to whatever logic selects the first agent (implicitly assuming that they will all have the same device+framework). This would fix the rewriting of x0 `len(network.agents())` times.

Regarding your suggestion @Simpag , I think we can definitely offer this logic for initialization (or a decorator) like we talked by email. Of course we will caution the users that it only works with `x` states.

Simpag replied:
Yes, I'm sorry I missed a parameter to my example. We would either have it inside the init loop:
```python
self.x0 = default_initialization(self.x0, i) # i is an agent
```
or assume that all agents have the same cost function and have:
```python
self.x0 = default_initialization(self.x0, network)
```
where we then use `i = network.agents()[0]`. 

In both cases we could maybe have an extra optional parameter called `stack_to` or something similar which would take care of the ADMM case where we need to stack copies. As for the decorator, I dont think it will be of too much value as both cases would be a one-liner with the same parameters. There is no real way for us to infer the shape and which parameter (if we dont restrict it to only `x` but we dont have to) to apply this transformation to.

Lastly, regarding the rewriting, the `to_array()` function doesnt do anything if you are trying to convert the object to the same framework+device that it already is. I made it this way so that you'd rather overdo it and not leave random bugs while keeping performance.

Simpag commented on decent_bench/utils/progress_bar.py:
```diff
@@ -1,17 +1,72 @@
+        else:
+            trial = (
+                task.fields["fields"].get("trial", "?") if "fields" in task.fields else task.fields.get("trial", "?")
+            )
```
I only tested it on windows and linux but usually macos is inline with linux

nicola-bastianello commented on decent_bench/utils/progress_bar.py:
```diff
@@ -1,45 +1,127 @@
 from dataclasses import dataclass
```
I see, thanks for the explanation. maybe I would remove "Trial" before the counter as this was what created confusion to me. it could be changed to "active trials", or better yet, we could add an explanation row above the progress bar. something where the meaning of each column is explained (algorithm, progress, total time, num. active trials).

nicola-bastianello replied:
is there a way to align the headers with the corresponding columns? this would improve readability quite a lot. also: I would change "Trial [x/y]" to just "x/y". I would also remove "Done/Total" and consider the % as part of the progress bar (since it's pretty easy to interpret)

Maybe it's possible to go to new line in the headers, so that the columns don't need to be spaced out while still ensuring that the headers are above the corresponding columns

sorry to be so picky..

Simpag commented:
Fixed a bug where the algorithms would fail if you stack the initialization, like in ADMM, while running async and more than 1 trial. This would fail because it got stacked the first trial and the second trial would already have the stacked initial variable which is not of correct shape (agent.cost.shape). Also fixed a bug where the progress bar sometimes does not finish before the framework tries to create a loading bar for calculating the statistics which would cause an error. The solution to this also fixed a bug where the progress bar sometimes left multiple copies of itself in the console.
